### DISCUSSION TITLE
Module updates from Terraform Cloud Registry

### DISCUSSION THREAD

**Discussion by joonvena** [renovatebot/renovate#18301] [Request Help] [answered]
### What would you like Renovate to be able to do?

It would be nice if Renovate could support updating Terraform modules from Terraform Cloud Registry.

### If you have any ideas on how this should be implemented, please tell us here.

Terrafrom Cloud provides Registry Module API to interact with the registry so it should be pretty straight forward to implement this using it.

https://www.terraform.io/cloud-docs/api-docs/private-registry/modules

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by viceice** [comment, marked as answer]
did you tried to override registry url? i though it should work. 🤔

**Comment by joonvena** [comment]
It did actually work! Thank for the tip 👍

**Reply by mdiianni** [reply]
Hello @joonvena @viceice 
Could you please provide an example of the snippet you used to override the configuration?
I am trying to fine tune a Renovate Kubernetes manifest to automate TF modules updates and patching pulling from both public and private registries, and the private registry hosted on Terraform Cloud.
So far I am trying with something like this for the `config.json`:
```json
{
  "platform": "gitlab",
  "endpoint": "https://gitlab.com/api/v4",
  "autodiscover": "false",
  "dryRun": false,
  "printConfig": true,
  "logLevel": "debug",
  "repositories": [
    "my-org/some-project/iac",
  ],
  "hostRules": [
    {
      "hostType": "terraform",
      "matchHost": "app.terraform.io",
      "token": "TERRAFORM_IO_TOKEN",
      "concurrentRequestLimit": 1
    }
  ],
  "packageRules": [
    {
      "matchDatasources": [
        "terraform-module"
      ],
      "registryUrls": [
        "https://registry.terraform.io/v1/modules/",
        "https://app.terraform.io/api/registry/v1/modules/my-org/"
      ]
    }
  ]
}
```

But I can only get the providers updated, 
Thanks

**Reply by joonvena** [reply]
Hi @mdiianni i only have this in my config.js and it works. I think there might be something wrong with those registryUrls there. Could you try just with the hostRules block and see if it works:

```
hostRules: [
     {
          matchHost: 'app.terraform.io',
          token: process.env.TF_CLOUD_TOKEN
     },
],
```

**Reply by mdiianni** [reply]
Thanks for the reply @joonvena ,

I tried that initially to get the public modules updated:

For example, this RDS module:
```
module "rds" {
  source  = "terraform-aws-modules/rds/aws"
  version = "3.2.0"

  [SOME MORE STUFF HERE]
}
```
Gets updated fine and it is of the form `source  = "terraform-aws-modules/...` 

But private modules like the following are not getting updated. Note the source is of the form: `source  = "app.terraform.io/my-org/...`
```
module "baseline" {
  source  = "app.terraform.io/my-org/baseline/tlz"
  version = "0.4.3"

  [SOME OTHER STUFF HERE]
}
```
I can also confirm the token is properly setup and has access to all the modules I want Renovate to "monitor".

Also when checking the logs it says the repository has been processed correctly with no warnings nor errors.

**Reply by mdiianni** [reply]
@secustor @joonvena do you think the issue I am having could be related to the Terraform version I am using (0.13), as mentioned in the comment above I cannot find a way to update the module version when it comes from a source like this: `source  = "app.terraform.io/my-org/baseline/tlz"`

**Reply by joonvena** [reply]
I don't think the Terraform version should matter at this stage. Have you tested out if you change the Terraform Cloud token to incorrect value does Renovate throw you some error or report something to Dependency Dashboard (If you are using them.). Mainly to check that does it actually try to even connect to the Terraform Cloud Registry.

**Reply by secustor** [reply]
@mdiianni Please create a new discussion and add your logs for the repo. Without logs it is difficult to help you.
